Add support for playing an InsufficientFunds warning

abcdefg30 opened the pull request:
Also enables those warnings in ts and ra.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Player/InsufficientFundsWarning.cs:
```diff
@@ -0,0 +1,57 @@
+		readonly InsufficientFundsWarningInfo info;
+		readonly PlayerResources resources;
+
+		int nextWarnTime = 0;
```
This needs to be set to `info.WarnInterval` in the ctor otherwise the condition checking this will be true immediately.

pchote commented:
I'm pretty sure that the original games didn't set this up as a repeating warning, which is going to be quite annoying for people who play the game efficiently (running with a very low float).

MustaphaTR commented:
In orginal game InsufficientFunds plays at two times, when you go out of credits, and when you try to build something when you are out of credits.  They doesn't repeat it unless you gain some credits. and lost it again iirc.

abcdefg30 commented:
Updated. It plays now when you reach zero money while trying to produce something.

Mailaender commented on OpenRA.Mods.Common/Traits/Player/InsufficientFundsWarning.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Provides the player with an audible warning when their storage is nearing full.")]
```
Copy pasta error.

abcdefg30 replied:
Ops, will fix.

abcdefg30 commented:
Fixed the typo.

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -82,6 +82,11 @@ public interface IProductionIconOverlay
 		bool IsOverlayActive(ActorInfo ai);
 	}
 
+	public interface INotifyInsufficientFunds
```
Please add `[RequireExplicitImplementation]`

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -82,6 +82,11 @@ public interface IProductionIconOverlay
 
+	public interface INotifyInsufficientFunds
+	{
+		void InsufficientFunds();
```
This should accept an `Actor self` parameter for consistency with other interfaces.

abcdefg30 commented:
Updated.

pchote commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -745,4 +746,4 @@ cd "$(SolutionDir)"</PostBuildEvent>
   </Target>
   -->
-</Project>
+</Project>
```
Files should end with newlines.  Please drop this change.

abcdefg30 commented:
Fixed again.

Unit158 commented:
This in d2k as well.

Mailaender commented:
> I'm pretty sure that the original games didn't set this up as a repeating warning, which is going to be quite annoying for people who play the game efficiently (running with a very low float).

This still overlaps with "Building". I also think it should replace it with "Insufficient Funds" and don't repeat the message.

Mailaender commented:
That is more like the original now and less annoying. :white_check_mark:

obrakmann commented:
I see two problems with this:

1. If you're building from multiple queues at the same time, you get a notification from each queue.
2. If you queue other items while production is ongoing, you no longer get the queued notification.

abcdefg30 commented:
Sorry guys, this wasn't actually completed yet (which is why I didn't leave an "Updated" comment).

I already had a small talk with @pchote on IRC to fix the issue @obrakmann pointed out.
Seems like I have to get back to interfaces for it, sorry @Mailaender,

abcdefg30 committed:
Add support for playing an InsufficientFunds warning

abcdefg30 committed:
Fix the horrible formatting in CustomBuildTimeValueExts

abcdefg30 committed:
Make use of the new InsufficientFunds warning in ra, d2k and ts

abcdefg30 commented:
Updated and rebased.

> 2. If you queue other items while production is ongoing, you no longer get the queued notification.

You'll get the notification now when it actually starts buildings. This is a side effect from fixing "This still overlaps with "Building!"" (https://github.com/OpenRA/OpenRA/pull/10570#issuecomment-174271390). Not sure if we want to call that an improvement. If it is unwanted, I'll just drop the change entirely and then we got the overlapping (except someone comes up with a good idea).

obrakmann commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -458,7 +471,23 @@ public void Tick(PlayerResources pr)
+			{
+				insufficientFundsPlayed = false;
+				foreach (var funds in insufficientFunds)
+					funds.SufficientFunds(owner.PlayerActor);
```
You need to move the `foreach` out of this block. Otherwise you won't get a warning in all circumstances. Ie. if you have $300 left and try to build a helipad, you'll get a warning. If you then build a E1 the build succeeds, which should reset the warning. However, if you try to build an E3 after the E1, the warning will not appear. Moving the `foreach` will fix that.

abcdefg30 commented:
Updated.

obrakmann commented:
:+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/16ccc0d5e736b44aec0e83a4eb1a131c1c2f8a68)

richlieb commented:
InsufficientFunds warning can be triggered too rapidly in the current playtest version of Dune (OpenRA-playtest-20160403).

For example your funds are at 100, your harvester is restocking your funds and you produce a building, a unit and a vehicle. The harvester raises your funds above 100 and in the next moment your funds get reduced below 100 due to your production and every time your funds drop below 100 the notifications gets triggered

As a result the notification sound overlaps, creates a kind of unpleasent acoustic canon.
